src/crimson/os/seastore/segment_cleaner.cc

```diff
@@ -448,30 +533,56 @@ SegmentCleaner::init_segments_ret SegmentCleaner::init_segments() {
 	auto segment_id = it.first;
 	return scanner->read_segment_header(
 	  segment_id
-	).safe_then([&segment_set, segment_id, this](auto header) {
-	  if (header.out_of_line) {
-	    logger().debug(
-	      "ExtentReader::init_segments: out-of-line segment {}",
-	      segment_id);
-	    init_mark_segment_closed(
-	      segment_id,
-	      header.journal_segment_seq,
-	      true);
-	  } else {
-	    logger().debug(
-	      "ExtentReader::init_segments: journal segment {}",
-	      segment_id);
-	    segment_set.emplace_back(std::make_pair(segment_id, std::move(header)));
-	  }
-	  return seastar::now();
+	).safe_then([&segment_set, segment_id, this](auto header)
+	  -> ExtentReader::scan_extents_ertr::future<> {
+	  return scanner->read_segment_tail(
+	    segment_id
+	  ).safe_then([this, segment_id, &segment_set, header](auto tail)
+	    -> ExtentReader::read_segment_tail_ertr::future<> {
+	    if (tail.segment_nonce != header.segment_nonce) {
```

Comment:
I don't think an invalid tail means the entire segment is invalid. IIUC currently only rolling will write down the tail. And it is impossible to write segment tail under unexpected power failure, but the segment can still contain valid records.

Reply:
We are not treating invalid tail as invalid segment, it just means we'll need to scan the segment's extents. See the code below.